### ISSUE TITLE
Today date cannot be selected from the emTripLog's date picker.

### ISSUE THREAD

**Issue by asiripanich** [e-mission/e-mission-docs#531]
A user reported that today date (3rd of May) couldn't be selected on emTripLog. My two test phones (iOS and Android) never have any issue with the date picker so I'm quite surprised to hear about this issue. 

![Screenshot_20200503-144058_emTripLog 1  copy](https://user-images.githubusercontent.com/17020181/80913679-cdbd6980-8d89-11ea-8254-bc4376640fba.jpg)

**Comment by shankari** [comment]
@asiripanich the screenshot alone doesn't tell me anything. I can see that the date is not selected, but have no clues as to **why** it is not selected. Are there any logs from the time that this occured?

**Comment by asiripanich** [comment]
@shankari Whoops sorry I thought I had sent you the log file for this.

**Comment by shankari** [comment]
so actually the logs were less helpful than I had hoped, but I think I figured out the issue. 

When `$scope.pickDay` is called, it passes `$scope.datepickerObject` into the datepicker. But `$scope.datepickerObject` is created when the controller is first created. So if a user launched the app on Sunday, didn't manually kill it, and the system didn't kill it, then when they resumed the app on Monday, the datepickerObject will still be the one from Sunday.

To verify if this is the case, kill the app and relaunch it.

If that fixes the problem, then we know this is the problem.
The fix is to recreate the `datepickerObject` on resume as well using 
https://ionicframework.com/docs/v1/api/service/$ionicPlatform/

so something like:

```
$ionicPlatform.on("resume", function() {
    // recreate date picker object
});
```

- Please refactor the code to create the datepicker instead of just copy-pasting it
- if this work, please submit a fix to master as well

**Comment by shankari** [comment]
I've attached a potential fix. This is **untested**. You can test it and if it works, please submit it as a PR to master.

[datepickerfix.patch.gz](https://github.com/e-mission/e-mission-docs/files/4570985/datepickerfix.patch.gz)

**Comment by asiripanich** [comment]
@atton16 could you test the patch please.

**Comment by asiripanich** [comment]
@shankari the user did relaunch the emTripLog app on Sunday.

**Comment by shankari** [comment]
> the user did relaunch the emTripLog app on Sunday.

@asiripanich not sure how that helps debug. The big question is whether the user killed the app before resuming it or not.

**Comment by asiripanich** [comment]
If the user relaunched the app, it means he had killed the app before opened it again right?

**Comment by shankari** [comment]
not necessarily. the app can be suspended (e.g. not in the foreground), and android may not kill it if there is no memory pressure. Then the user can reselect the app from the task list and it will resume with the previous state intact.

**Comment by asiripanich** [comment]
Huh, I always thought the closing an app by swiping the app's window up would kill it.

**Comment by shankari** [comment]
that will kill it, but is that what the user did? Or did they just launch another app instead? Imprecise descriptions make it hard to debug what is going on. Please list out the actual steps that the user performed - e.g. "swiped up app's window", "selected app from home screen" instead of "relaunched"

**Comment by asiripanich** [comment]
Sorry. That is what I meant by relaunching. The user swiped up emTripLog app's window and selected the app from home screen.

**Comment by shankari** [comment]
and even after swiping up the window and relaunching the app from the home screen, they still could not select the correct day?

Can you or they try to reproduce again? Because if the problem persists even after a force-kill + relaunch from the home screen, then the proposed fix won't work.

To reproduce:
- launch the app today
- do not force kill - leave it running in the background for a whole day
- bring it to the foreground tomorrow
- you won't be able to select today's date
- now, force kill the app
- relaunch it from the home screen
- now you should be able to select today's date

If the scenario above is reproducible, then I know the problem and the patch will work. Otherwise, the issue needs more investigation.

Please report the results of the experiment in a detailed, step by step manner as outlined above.

**Comment by asiripanich** [comment]
In the initial email you suggested to kill and reopen the app and that worked for the user. That was on a samsung S9. I’m going to try and to reproduce that on an iPhone and an android Pixel 2.

**Comment by shankari** [comment]
ok, given this

> You suggested to kill and reopen the app and that worked for the user.

I'm going to start making the changes assuming that this is the problem 😄

**Comment by asiripanich** [comment]
Will do. Thank you so much Shankari.

**Comment by asiripanich** [comment]
> and even after swiping up the window and relaunching the app from the home screen, they still could not select the correct day?
> 
> Can you or they try to reproduce again? Because if the problem persists even after a force-kill + relaunch from the home screen, then the proposed fix won't work.
> 
> To reproduce:
> 
>     * launch the app today
> 
>     * do not force kill - leave it running in the background for a whole day
> 
>     * bring it to the foreground tomorrow
> 
>     * you won't be able to select today's date
> 
>     * now, force kill the app
> 
>     * relaunch it from the home screen
> 
>     * now you should be able to select today's date
> 
> 
> If the scenario above is reproducible, then I know the problem and the patch will work. Otherwise, the issue needs more investigation.
> 
> Please report the results of the experiment in a detailed, step by step manner as outlined above.

**Summary**: Tested this on two phones and could not reproduce the problem.

**Devices**: iPhone X iOS 13.3.1, Google Pixel 2 Android 10

**Steps**
- Relaunched the app
- Leave it running it the background
- Launched the app the next day again. The app reloaded and the date appeared correctly on both phones.

**Comment by asiripanich** [comment]
Just to be clear, when you said to launch the app that doesn't mean to do a fresh install and sign in right?

**Comment by shankari** [comment]
No, I meant click on the icon in the home screen.
Essentially, if the app is not killed, but only suspended, then the date picker is cached and reused.


Note that the date will show up correctly, but if you open the datepicker, then it won't allow you to select the current date.

I'm going to try on my own personal test phone.

**Comment by asiripanich** [comment]
> No, I meant click on the icon in the home screen.
> Essentially, if the app is not killed, but only suspended, then the date picker is cached and reused.
> 
> Note that the date will show up correctly, but if you open the datepicker, then it won't allow you to select the current date.
> 
> I'm going to try on my own personal test phone.

For me, the date pickers displayed correctly on both phones.

**Comment by asiripanich** [comment]
@shankari do we have any progress on this? Thanks.

**Comment by shankari** [comment]
@asiripanich I have created the branch to work on it 😄 I am sure you realize that I have other demands on my time as well.

**Comment by shankari** [comment]
I tried to reproduce this on my personal phone, but got the same behavior as @asiripanich 
But I also noticed that when I clicked on the app from the home screen, it went through the full relaunch process. Next time, I will select it from the task list.

**Comment by shankari** [comment]
ok I can reproduce this in the emulator and @asiripanich you should be able to reproduce on your test phone as well. Without a reproducible test, we can't figure out whether the fix actually works. The test is as important as the fix!

To reproduce:
- launch the app and open the datepicker
- close the datepicker
- launch the settings app WITHOUT killing emTripLog
- change the date to a few days later
- go to the task list and select emTripLog (DO NOT launch from home screen)
- the current (fake) date is not selectable

**Comment by shankari** [comment]
Using this reproduction, I was able to verify that the issue is fixed and we can select all dates upto the present. However, "today" is still set to the time that the datepicker was initialized. This appears to be set in the datepicker during creation.

https://github.com/rajeshwarpatlolla/ionic-datepicker/blob/master/src/ionic-datepicker.provider.js#L29

Need to figure out how to refresh that, or need to fork

**Comment by asiripanich** [comment]
> I tried to reproduce this on my personal phone, but got the same behavior as @asiripanich 
> 
> But I also noticed that when I clicked on the app from the home screen, it went through the full relaunch process. Next time, I will select it from the task list.

I did try that on my android but the screen also went through the full relaunch process.

**Comment by asiripanich** [comment]
> @asiripanich I have created the branch to work on it 😄 I am sure you realize that I have other demands on my time as well.

Of course that is why I asked to see where @atton16 and I can help. :)

**Comment by asiripanich** [comment]
> ok I can reproduce this in the emulator and @asiripanich you should be able to reproduce on your test phone as well. Without a reproducible test, we can't figure out whether the fix actually works. The test is as important as the fix!
> 
> 
> 
> To reproduce:
> 
> - launch the app and open the datepicker
> 
> - close the datepicker
> 
> - launch the settings app WITHOUT killing emTripLog
> 
> - change the date to a few days later
> 
> - go to the task list and select emTripLog (DO NOT launch from home screen)
> 
> - the current (fake) date is not selectable
> 
> 

Um but this doesn’t seem like a normal case... the user who reported this didnt mention that the user had changed the date in the phone setting but I remember seeing in the user’s log that the app was reloaded a few times by pulling down on the diary screen.

**Comment by shankari** [comment]
in that case, they didn't change the date - they just waited until the date changed normally. that is a very slow process for a reproducible test though, so this is a way to speed up testing.

**Comment by shankari** [comment]
@asiripanich try it out and ensure that you can reproduce it so that you can ensure that the fix works.

**Comment by asiripanich** [comment]
> in that case, they didn't change the date - they just waited until the date changed normally. that is a very slow process for a reproducible test though, so this is a way to speed up testing.

Yes I can reproduce that on my android test phone.

**Comment by shankari** [comment]
Fixed in #667, tested in #668
